87.md

```diff
@@ -0,0 +1,129 @@
+
+## Rationale
+
+Nostr's discoverability and transparent event interaction is one of its most interesting/novel mechanics.
+This NIP provides a simple way for users to discover ecash mints recommended by other users and to interact with them.
+
+### Parties involved
+
+There are three actors to this workflow:
+
+* An ecash mint operator, announces their mint and its capabilities.
+    * Publishes `kind:38173` or `kind:38172`, detailing how to connect to it and its capabilities.
+* user A, who recommends an ecash mint
+    * Publishes `kind:18173`
+* user B, who seeks a recommendation for an ecash mint
+    * Queries for `kind:18173` and, based on results, queries for `kind:38173`/`kind:38172`
+
+## Events
+
+### Recommendation event
+```json
+{
+  "kind": 18173,
+  "pubkey": <recommender-user-pubkey>,
+  "tags": [
+    ["u", <recommended-cashu-mint-url>, "cashu"],
+    ["u", <recommended-fedimint-invite-code>, "fedimint"],
+    ["u", <recommended-cashu-mint-url>, "cashu"],
+    ["a", "38173:fedimint-pubkey:<d-identifier>", "wss://relay1", "fedimint"],
+    ["a", "38172:cashu-mint-pubkey:<d-identifier>", "wss://relay2", "cashu"]
+  ],
+  "content": ""
+}
+```
+
+`u` tags are used to point directly to the ecash mint.
+The first value of the tag is the URL or invite code of the ecash mint, and the second value is the type of ecash mint (currently `cashu` or `fedimint`).
+Multiple `u` tags can appear on the same `kind:18173`.
+
+`a` tags are used to point to the `kind:38173`/`kind:38172` event of the ecash mint.
+The first value of the tag is the `kind:38173`/`kind:38172` event identifier, the second value of the tag is a relay hint, and the third value of the tag is the type of ecash mint (currently `cashu` or `fedimint`).
+
+## Ecash Mint Information
+
+Cashu mints SHOULD publish `kind:38172` events to announce their capabilities and how to connect to them.
+
+For cashu mints, the `u` tag SHOULD be the URL to the cashu mint and it should list the nuts that the cashu mint supports.
+
+The `d` tag SHOULD be the mint's pubkey (found when querying `/v1/info`), this way users can query by pubkey and find the mint announcement.
+
+```json
+{
+  "kind": 38172,
+  "pubkey": "<application-pubkey>",
+  "content": "<optional-kind:0-style-metadata>",
+  "tags": [
+    ["d", <cashu mint pubkey>],
+    ["u", "https://cashu.example.com"],
+    ["nuts", "1,2,3,4,5,6,7"]
+  ]
+}
+```
+
+Fedimints SHOULD publish `kind:38173` events to announce their capabilities and how to connect to them.
+
+For fedimints, it should list all known fedimint invite codes in `u` tags and it should list the modules it supports.
+
+The `d` tag SHOULD be the federation id, this way users can query by federation id and find the fedimint announcement.
+
+```json
+{
+  "kind": 38173,
+  "pubkey": "<application-pubkey>",
+  "content": "<optional-kind:0-style-metadata>",
+  "tags": [
+    ["d", <federation-id>],
+    ["u", "fed11abc.."],
+    ["u", "fed11xyz.."],
```

Comment:
An invite code is just a bech32 encoded federation ID and a list of API urls. The proposed encoding is thus highly redundant: each invite contains the federation id again and possibly overlapping API URLs. Instead I'd recommend to go with:

```
    ["d", <federation-id>],
    ["u", "wss://guardian-1.url"],
    ["u", "wss://guardian-2.url"],
```

We do include the guardian id for each URL in the invite, but I don't think it's strictly necessary. If we want full future feature-parity (if that ID ever gets used for anything) we could encode the guardian id in the URL announcement.

```
    ["d", <federation-id>],
    ["u", "1,wss://guardian-1.url"],
    ["u", "2,wss://guardian-2.url"],
```

To decode an invite code use `fedimint-cli dev decode-invite-code`:

```
$ fedimint-cli dev decode-invite-code fed11qgqpw9thwvaz7te3xgmjuvpwxqhrzw33xvenvve0qyqjpq7rkttxqknn0p30lsyxnljwqw0ff2l55t7rc6xxgup8zfguc9dhh9d3yd
{
  "url": "ws://127.0.0.1:13363/",
  "federation_id": "83c3b2d6605a737862ffc0869fe4e039e94abf4a2fc3c68c6470271251cc15b7"
}
```

Our support for multiple URLs in dev tooling may still be a bit lacking support for it was added recently. Currently you'd call `fedimint-cli invite-code <guardian-id>` for each guardian and then decode the invite, getting their API URL.

Reply:
It seems like we'd need the index to build the invite code

https://github.com/fedimint/fedimint/blob/ca6176c76de0e92c68392e82b4d8ab1a7c7df87b/fedimint-core/src/api.rs#L827

Reply:
I think this makes sense to me, my only concern would be if there would ever be changes to the encoding / extra data added to the invite code. Then we'd have to change this encoding here as well. Encoding the invite code gives us better guarantees of backwards compatibility